### PR TITLE
Make sure historical state events don't come down incremental `/sync` (MSC2716)

### PR DESCRIPTION
**Pull request by MadLittleMods**
Part of MSC2716: https://github.com/matrix-org/matrix-doc/pull/2716

Related to https://github.com/matrix-org/synapse/issues/11241

Split off from https://github.com/matrix-org/complement/pull/221

### CODE REVIEW

File: tests/msc2716_test.go [hunk 1]
```diff
@@ -248,19 +252,20 @@ func TestImportHistoricalMessages(t *testing.T) {
...
-			// between, we're probably safe to assume it won't sync
-			alice.SyncUntil(t, "", `{ "room": { "timeline": { "limit": 3 } } }`, "rooms.join."+client.GjsonEscape(roomID)+".timeline.events", func(r gjson.Result) bool {
-				if r.Get("event_id").Str == historicalEventId {
-					t.Fatalf("We should not see the %s historical event in /sync response but it was present", historicalEventId)
+			// Sync from before we did any batch sending until we find the
+			// eventIDAfterHistoricalImport. If we're able to see
+			// eventIDAfterHistoricalImport without any the
+			// historicalEventIDs/historicalStateEventIDs in between, we're probably
+			// safe to assume it won't sync.
+			alice.SyncUntil(t, since, "", "rooms.join."+client.GjsonEscape(roomID)+".timeline.events", func(r gjson.Result) bool {
```
Comment (MadLittleMods) [inline tests/msc2716_test.go:266]: Made this test more clear on what's happening. We now paginate sync from before we `/batch_send` until an event that occurred after `/batch_send` and just make sure we didnt' see any historical state or events in between.

File: tests/msc2716_test.go [hunk 2]
```diff
@@ -248,19 +252,20 @@ func TestImportHistoricalMessages(t *testing.T) {
...
+				if includes(r.Get("event_id").Str, historicalEventIDs) || includes(r.Get("event_id").Str, historicalStateEventIDs) {
```
Comment (MadLittleMods) [inline tests/msc2716_test.go:267]: Here is where we additionally check for the `historicalStateEventIDs` now

File: tests/msc2716_test.go [hunk 3]
```diff
@@ -220,7 +220,7 @@ func TestImportHistoricalMessages(t *testing.T) {
 			validateBatchSendRes(t, as, roomID, batchSendRes, false)
 		})
 
-		t.Run("Historical events from /batch_send do not come down in an incremental sync", func(t *testing.T) {
+		t.Run("Historical events from batch_send do not come down in an incremental sync", func(t *testing.T) {
```
Comment (MadLittleMods) [inline tests/msc2716_test.go:223]: Removing the `/` in the test name so I can actually run it individually, `COMPLEMENT_ALWAYS_PRINT_SERVER_LOGS=1 COMPLEMENT_DIR=../complement ./scripts-dev/complement.sh TestImportHistoricalMessages/parallel/Historical_events_from_batch_send_do_not_come_down_in_an_incremental_sync`

See https://gist.github.com/MadLittleMods/4ab08f51609fab759247f299a4e33406 for why there is a problem using a `/` to match a single test.

**Review by kegsay — APPROVED** [review]

**Comment by MadLittleMods** [comment]
Thanks for the review @kegsay 🐿

**Pull request merged**